File: Backend/API/Entities/Basket.cs
using System;

namespace API.Entities;

public class Basket
{
    public Guid Id { get; set; }
    public required string BasketId { get; set; }
    public List<BasketItem> Items { get; set; } = [];

    public void AddItem(Product product, int quantity)
    {
        if (product is null)
            ArgumentNullException.ThrowIfNull(product);
        if (quantity <= 0)
            throw new ArgumentException("Quantity should be greater than zero", nameof(quantity));

        BasketItem? existingItem = FindItem(product.Id);

        if (existingItem is null)
        {
            Items.Add(new() { Product = product, Quantity = quantity });
        }
        else
        {
            existingItem.Quantity += quantity;
        }
    }

    public void RemoveItem(Guid productId, int quantity)
    {
        if (quantity <= 0)
            throw new ArgumentException("Quantity should be greater than zero", nameof(quantity));

        BasketItem? item = FindItem(productId);
        if (item is null)
            return;

        item.Quantity -= quantity;
        if (item.Quantity <= 0)
            Items.Remove(item);
    }

    private BasketItem? FindItem(Guid productId)
    {
        return Items.FirstOrDefault(item => item.ProductId == productId);
    }
}

File: Frontend/src/models/Basket.ts
import { Product } from "./Product";

export type Basket = {
  basketId: string;
  items: Item[];
};

export class Item {
  constructor(product: Product, quantity: number) {
    this.brand = product.brand;
    this.description = product.description;
    this.name = product.name;
    this.pictureUrl = product.pictureUrl;
    this.price = product.price;
    this.productId = product.id;
    this.quantity = quantity;
    this.quantityInStock = product.quantityInStock;
    this.type = product.type;
  }

  productId: string;
  quantity: number;
  name: string;
  description: string;
  price: number;
  pictureUrl: string;
  type: string;
  brand: string;
  quantityInStock: number;
}
